### ISSUE TITLE
Type argument inference does not take call's contextual type into account

### ISSUE THREAD

**Issue by zakjan** [Microsoft/TypeScript#10245]
**TypeScript Version:**  2.0.0

**Code**

``` ts
function map(array: number[]): [number, number][] {
    return array.map((x: number) => [x, x * x]);
}
```

**Expected behavior:**
no error

**Actual behavior:** 

```
error TS2322: Type 'number[][]' is not assignable to type '[number, number][]'.
  Type 'number[]' is not assignable to type '[number, number]'.
    Property '0' is missing in type 'number[]'.
```

**Comment by DanielRosenwasser** [comment]
What's happening here is that the call to `array.map` doesn't try to infer its return type from its contextual type. It only tries to infer from its arguments.

A temporary workaround would be to give `[number, number]` as a type argument to `map`:

```ts
function map(array: number[]) {
    return array.map<[number, number]>(x => [x, x * x]);
}
```

**Comment by whiteblackkeys** [comment]
Here some other examples:

`export interface DataModel {
    prop1: string;
    prop2: number;
}

const array = [1,2,3,4];

// NO ERROR
const datamodels2: DataModel[] = array.map(i => ({
    prop1: 'asdf',
    prop2: 1234,
    extra: 'extra'
}))

// NO ERROR
const datamodels4: DataModel[] = array.map(i => {
    return {
        prop1: 'asdf',
        prop2: 1234,
        extra: 'extra'
    } 
});

// NO ERROR
const datamodels5: DataModel[] = array.map(i => {
    return {
        prop1: 'asdf',
        prop2: 1234,
        extra: 'extra'
    }  as DataModel
});

// NO ERROR
const datamodels6: DataModel[] = array.map(i => ({
        prop1: 'asdf',
        prop2: 1234,
        extra: 'extra'
}) as DataModel);
`

**Comment by RyanCavanaugh** [comment]
@whiteblackkeys Unrelated. Root cause is #241 - function return expression widening causes the object literal freshness to be lost, causing excess properties to not be checked.

**Comment by ahejlsberg** [comment]
Fixed by #29478.
